### DISCUSSION TITLE
[Dynamic Links] How to get query parameters from dynamic link?

### DISCUSSION THREAD

**Discussion by achubutkin** [firebase/flutterfire#5579] [Feature request]
**Describe the bug**
The documentation does not describe how to get a dynamic link URL or its query parameters. Attention! Dynamic link query parameters, not included deeplink!

**Comment by karapetyan** [comment]
Same here. How to get query parameters from firebase dynamic link inside app? 

https://testapp.page.link/ios?param=paramValue
How to get "paramValue"

**Reply by SnowStar0423** [reply]
The created Deep Link does not give the query Parameters when user enters it in the Chrome browser.
In other words, Chrome browser removes the generated query parameters when my friend clicks the link.
What's why?
Have you received the correct parameters values from the created Deep Links again?

**Comment by mahmoudelgawad** [comment]
Same here,
```
 final data = await FirebaseDynamicLinks.instance.getInitialLink();
 final Uri deepLink = data?.link;
```
when `deepLink .hasQuery` is always `false`

**Reply by SnowStar0423** [reply]
The created Deep Link does not give the query Parameters when user enters it in the Chrome browser.
In other words, Chrome browser removes the generated query parameters when my friend clicks the link.
What's why?
Have you received the correct parameters values from the created Deep Links again?

**Comment by mahmoudelgawad** [comment]
I fixed by the following document

https://firebase.google.com/docs/dynamic-links/create-manually

**Reply by SnowStar0423** [reply]
The created Deep Link does not give the query Parameters when user enters it in the Chrome browser.
In other words, Chrome browser removes the generated query parameters when my friend clicks the link.
What's why?
Have you received the correct parameters values from the created Deep Links again?

**Comment by KlausJokisuo** [comment]
Any news on this?

**Comment by KingIdee** [comment]
You can do this. Where your query param is `code` :

```
final Uri deepLink = dynamicLink?.link;
deepLink.queryParameters['code']
```

**Comment by luis-cruzt** [comment]
The approach of @KingIdee is the correct one.

Example, following @karapetyan question:

```
Future handleDynamicLinks() async {

    // Get initial dynamic link if the app is started using the link
    final PendingDynamicLinkData data = await FirebaseDynamicLinks.instance.getInitialLink();
    
    handleDeepLink(data);

    // Into foreground from dynamic link logic
    FirebaseDynamicLinks.instance.onLink(
      onSuccess: (PendingDynamicLinkData dynamicLinkData) async {
        handleDeepLink(dynamicLinkData);
      },
      onError: (OnLinkErrorException e){
        print('Dynamic Link Failed: $e');
      }
   );
}

void handleDeepLink(PendingDynamicLinkData data) {

         final Uri deepLink = data?.link;
         var paramValue = deepLink.queryParameters['paramValue'];

         if (deepLink != null) {
           print('deepLink detected: $deepLink');
           print('deepLink data: $paramValue');
              
         }         
 }
```

**Reply by SnowStar0423** [reply]
This is how to get the value of the query parameters when they are in the regenerated deep links.
In fact, when we create a new dynamic link programmatically, if we type the created deep links or click it on other app, we don't receive the query parameters from the deep links.
Did you receive the value of the query parameters in your project really?

**Comment by hackerunet** [comment]
is it possible to pass dynamic query parameters instead of plain parameters? normally an application creates a token from backend and send a link to the email with a token, how can I pass that token to the dynamic link? right now I can get a query parameter if is defined in the console with ?parameter but no matter if my link has a value, in the console as the parameter has no value at all I get null, if I add in the console ?parameter=xxx I get xxx but, that's not the idea. so, anyone can please help me out to understand this, since the documentation has no enough information regarding this.

**Comment by astronaute77** [comment]
I face the same issue, the query parameters are lost during URL conversion.
I opened another ticket highlighting the issue:
https://github.com/FirebaseExtended/flutterfire/issues/3125

**Comment by kostacasa** [comment]
It's really bizarre that the e-mail templating system from Firebase auth is essentially incompatible with dynamic links.

In the Firebase docs here:
https://firebase.google.com/docs/dynamic-links/create-manually

It says:

> Specify a URL that your app can handle, typically the app's content or payload, which initiates app-specific logic (such as crediting the user with a coupon or displaying a welcome screen). This link must be a well-formatted URL, **be properly URL-encoded**, use either HTTP or HTTPS, and cannot be another Dynamic Link.

Bolded section being key. 

However, Firebase e-mail templating attaches the mode and code query parameterrs `mode=<action>&oobCode=<code>` without URL encoding anything. That means it's impossible to define the password reset link as a dynamic link by using something like this:
```
https://your_subdomain.page.link/?link=https%3A%2F%2Fyour.domain%2F__%2Fauth%2Faction%2F%3F
```

Because firebase will tack on the query parameters without URLEncoding them, resulting in a link that looks like:
```
https://your_subdomain.page.link/?link=https%3A%2F%2Fyour.domain%2F__%2Fauth%2Faction%2F%3F&mode=resetPassword&oobCode=9icrng4K_sp_MSbYg2onUghNyYDROiX2EQLTaRdz4bofAAF0DeRWzw&lang=en
```
Notice the ampersand `&` characters. By inserting a non-urlencoded ampersand at the end, those parameters simply "escape" from the `link=` query parameter and become parameters of the page.link instead. 

Since the dynamic link system knows nothing about "oobCode" and other parameters, it will simply drop them, and the final link will miss the necessary context to complete the operation.

**Comment by brad353** [comment]
Any updates on this? I'm also facing the same issue.

**Comment by varunkumarmedam** [comment]
You can use https://pub.dev/packages/uni_links. This will help in getting the raw URL having query params also. So you don't need to use the flutter dynamic links package if you don't need the retrieving link data.

I know this is not the solution we really need, but it works at least for now

**Comment by astronaute77** [comment]
> 
> 
> You can use https://pub.dev/packages/uni_links. This will helps in getting the raw url having query params also. So you dont need to use flutter dynamic links package if you dont need the retrieving link data

Uni links is totally different use-case unrelated to this issue. This is about firebase dynamic links losing initial query parameters.

**Comment by varunkumarmedam** [comment]
> > You can use https://pub.dev/packages/uni_links. This will helps in getting the raw url having query params also. So you dont need to use flutter dynamic links package if you dont need the retrieving link data
> 
> Uni links is totally different use-case unrelated to this issue. This is about firebase dynamic links losing initial query parameters.

Of course, this is not related to Firebase Dynamic Links, I just explained how to make the things work. Uni_links helped my project to achieve what I need that which cant be done with Dynamic links

**Reply by LwinMaungMaungDev** [reply]
That's right, using uni_links can fetch the raw url including params, for example, "https://deeplinkname.page.link/?ABCD1234", where ABCD1234 is the parameter. Parameters should be a-z and numbers (cannot include special chars *, >, and &).

**Comment by astronaute77** [comment]
> 
> 
> > > You can use https://pub.dev/packages/uni_links. This will helps in getting the raw url having query params also. So you dont need to use flutter dynamic links package if you dont need the retrieving link data
> > 
> > 
> > Uni links is totally different use-case unrelated to this issue. This is about firebase dynamic links losing initial query parameters.
> 
> Of course, this is not related to Firebase Dynamic Links, I just explained how to make the things work. Uni_links helped my project to achieve what I need that which cant be done with Dynamic links

I’m happy it works for you an uni_links is great. It just is not a workaround for this issue you are commenting on.

**Comment by marcscherer** [comment]
Any updates on this?

**Comment by caitcif** [comment]
SOLVED for me anyway
I have spent hours on this issue and finally decided to create the link programmatically to see if it made a difference, and it did!

The link I created manually which did not work was
https://iomoney2.page.link/?link=https://iomoney.web.app/#/?report=EXAMPLE&ibi=iOmoneyLite

In iOS Swift, calling DynamicLinks.dynamicLinks().handleUniversalLink returned https://iomoney.web.app and would lose my report=EXAMPLE parameter

When I create the link programmatically I get:

https://iomoney2.page.link/?link=https%3A%2F%2Fiomoney%2Eweb%2Eapp%2F%23%2F%3Freport%3DEXAMPLE&ibi=iOmoneyLite

When I pass this to handleUniversalLink I get https://iomoney.web.app/#/?report=EXAMPLE which is what I want

After trying many combinations, it seems the # sign and the = sign between report and EXAMPLE are the only two characters throwing it off. If I use %23 for # and %3D for = it works again with this:

https://iomoney2.page.link/?link=https://iomoney.web.app/%23/?report%3DEXAMPLE&ibi=iOmoneyLite

**Comment by radiKal07** [comment]
The problem is inside Firebase REST API, but can be fixed from the plugin as well.
The plugin does this REST call internally:

```
curl --request POST \
  --url 'https://firebasedynamiclinks.googleapis.com/v1/reopenAttribution?key={firebase_key}' \
  --header 'content-type: application/json' \
  --header 'x-ios-bundle-identifier: {bundle_identifier}' \
  --data '{
	"bundle_id": "{bundle_identifier}",
	"sdk_version": "4.0.8",
	"requestedLink": "https://subdomain.domain.com/path?param=123456"
}'
```

This call returns:

```
{
  "deepLink": "https://subdomain.domain.com/path",
  "resolvedLink": "https://subdomain.domain.com/?link=https://subdomain.domain.com/path&apn={bundle_identifier}&isi=1493910790&ibi={bundle_identifier}&st={st}&sd={sd}?&si={si}"
}
```

Then the URL sent back to Flutter is the URL got from the response from the `deepLink` field.
Notice how in the request we sent `https://subdomain.domain.com/path?param=123456` and in the response we got back `https://subdomain.domain.com/path`. So Firebase backend removes `?param=123456`. What can the Flutter plugin do is ignore the `deepLink` from response and send back to Dart the original `requestedLink` which also has the params.

This is what I noticed for iOS.
On Android the URL does not have the query parameters from the start, so I don't think we can make it work just from the Flutter plugin.

@iapicca what do you think?

**Comment by radiKal07** [comment]
@varunkumarmedam using `uni_links` on Android gives you the option to open the URL either via Chrome or your own app. If user opens via your own app, the query parameters are present, if they open it via Chrome we again lose the query parameters, so it's not a reliable solution. Did you find a fix for this issue?

**Comment by varunkumarmedam** [comment]
@tudor07 thats a nice explanation I wanted to give. But fortunately `uni_links` also still not the best solution. But when I opened my app from whatsapp it works(dont works if app runs in background) its crazy but I too waiting for the perfect solution.

**Comment by caitcif** [comment]
I noticed that short links work around this issue but I have other issues with them so for the time being I am modifying the url string in

func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {

        var handled = false
        if userActivity.activityType == NSUserActivityTypeBrowsingWeb && userActivity.webpageURL != nil {
            let origStr = userActivity.webpageURL!.absoluteString
            // this is a hack for universal link to retain my parameters
            let url = URL(string: origStr.replacingOccurrences(of: "iomoney.web.app/#/?event=", with: "iomoney.web.app/%23/?event%3D"))
            if (url != nil) {
                handled = DynamicLinks.dynamicLinks().handleUniversalLink(url!) { (dynamicLink, error) in
                    if (error == nil && dynamicLink != nil && dynamicLink!.url != nil) {
                        self.open(url: dynamicLink!.url!)
                    }
                    else {
                        print(error!)
                    }
                }
            }
        }
        return handled
}

**Comment by varunkumarmedam** [comment]
@caitcif hope you know, we are talking about flutter firebase dynamic links here...

**Comment by caitcif** [comment]
> 
> 
> @caitcif hope you know, we are talking about flutter firebase dynamic links here...

OK, I don't know the difference but this thread started with firebase dynamic links and the exact issue I am seeing too.

**Comment by radiKal07** [comment]
@caitcif how did you fix for Android?

**Comment by caitcif** [comment]
> 
> 
> @caitcif how did you fix for Android?

I don't have an Android version, just iOS and Web. But if you can replace the # and ? characters in the dynamic link with their escape codes before passing to handleUniversalLink it should work. See my original post:

The link I created manually which did not work was
https://iomoney2.page.link/?link=https://iomoney.web.app/#/?report=EXAMPLE&ibi=iOmoneyLite

In iOS Swift, calling DynamicLinks.dynamicLinks().handleUniversalLink returned https://iomoney.web.app and would lose my report=EXAMPLE parameter

When I create the link programmatically I get:

https://iomoney2.page.link/?link=https%3A%2F%2Fiomoney%2Eweb%2Eapp%2F%23%2F%3Freport%3DEXAMPLE&ibi=iOmoneyLite

When I pass this to handleUniversalLink I get https://iomoney.web.app/#/?report=EXAMPLE which is what I want

After trying many combinations, it seems the # sign and the = sign between report and EXAMPLE are the only two characters throwing it off. If I use %23 for # and %3D for = it works again with this:

https://iomoney2.page.link/?link=https://iomoney.web.app/%23/?report%3DEXAMPLE&ibi=iOmoneyLite

**Comment by varunkumarmedam** [comment]
> > @caitcif how did you fix for Android?
> 
> I don't have an Android version, just iOS and Web. But if you can replace the # and ? characters in the dynamic link with their escape codes before passing to handleUniversalLink it should work. See my original post:
> 
> The link I created manually which did not work was
> https://iomoney2.page.link/?link=https://iomoney.web.app/#/?report=EXAMPLE&ibi=iOmoneyLite
> 
> In iOS Swift, calling DynamicLinks.dynamicLinks().handleUniversalLink returned https://iomoney.web.app and would lose my report=EXAMPLE parameter
> 
> When I create the link programmatically I get:
> 
> https://iomoney2.page.link/?link=https%3A%2F%2Fiomoney%2Eweb%2Eapp%2F%23%2F%3Freport%3DEXAMPLE&ibi=iOmoneyLite
> 
> When I pass this to handleUniversalLink I get https://iomoney.web.app/#/?report=EXAMPLE which is what I want
> 
> After trying many combinations, it seems the # sign and the = sign between report and EXAMPLE are the only two characters throwing it off. If I use %23 for # and %3D for = it works again with this:
> 
> https://iomoney2.page.link/?link=https://iomoney.web.app/%23/?report%3DEXAMPLE&ibi=iOmoneyLite

But when I tried it in android it seems not working for me. But happy if it works in iOS.

Did it worked in android for anyone..?!

**Comment by tareqdayya** [comment]
For everyone who's still looking for a way to do this:
- Every dynamic link has 2 forms: a short and a long one. You NEED to use the long one [check your dynamic link in firebase console]
- The long version of the dynamic link will look something like this:
`https://domain.page.link/?link=https://domain.com/coupon&apn=com.domain&efr=1`
- Parse the query params and re-construct the link such that your queries end up attached to the `link` query param. Example:
`https://domain.page.link/?link=https://domain.com/coupon?myquery=123&apn=com.domain&efr=1`
In other words, DO NOT DO THIS:
`https://domain.page.link/?link=https://domain.com/coupon&apn=com.domainl&efr=1&myquery=123`
- Send the link [including through Firebase Verification Emails]

I've tried this on web, and I can get the queries out of the url. I'm hopeful your library on ios and android can get the queries from the link as well.

**Reply by AllmaG3st** [reply]
This solution worked for me on IOS after some changes. (did not tested on Android because there is no such requirement)
1. We need long form as described, to get it go to your deep link details, it will look like this:
https://yourApp.page.link/?link=https://yourApp.com/&apn=com.youApp&isi=1111111111&ibi=com.yourApp.app&efr=1
2. add parameter 
3. change characters to receive something like this:
https://yourApp.page.link/?link=https%3A%2F%2FyourApp.com%2Fcourse%3FcourseId%3D123&apn=com.yourApp&isi=111111111&ibi=com.yourApp.app&efr=1

and now it works in my react native app and I receive url with parameter when using getInitialLink()

**Comment by larssn** [comment]
> * Parse the query params and re-construct the link such that your queries end up attached to the `link` query param. 

That's not always gonna work, sadly, if you are following an external redirect (like in most auth flows).

**Comment by tareqdayya** [comment]
> > * Parse the query params and re-construct the link such that your queries end up attached to the `link` query param.
> 
> That's not always gonna work, sadly, if you are following an external redirect (like in most auth flows).

Isn't this the job of the external redirector though? e.g. Configure the authentication provider to keep query params.
I don't think this is related to either the OP or even dynamic links in general.

**Comment by larssn** [comment]
Most auth flows have you set up a redirect url, that redirects back to the app. This part is static, as it is hosted by the external provider you want access to. They normally redirect to the given url, and then attach something like an access token as a query param.

My point was simply that if the above is the case, then it's not solvable with dynamic links as the query params are lost.

I ended up solving it by redirecting to a cloud function, which then grabs the query params, and then redirects to the dynamic link.

**Comment by astronaute77** [comment]
> 
> 
> > > * Parse the query params and re-construct the link such that your queries end up attached to the `link` query param.
> > 
> > 
> > That's not always gonna work, sadly, if you are following an external redirect (like in most auth flows).
> 
> Isn't this the job of the external redirector though? e.g. Configure the authentication provider to keep query params.
> I don't think this is related to either the OP or even dynamic links in general.

No, it is the job of dynamic links to not lose incoming query parameters.

Dynamic Link
https://monadeo.page.link/github_auth?code=xxxxx

Converts to:

Observed:
https://monadeo-test.firebaseapp.com/__/auth/handler

Expected:
https://monadeo-test.firebaseapp.com/__/auth/handler?code=xxxxx

Try yourself to implement Github authentication for example and you’ll see that it cannot work because of the above.

**Comment by varunkumarmedam** [comment]
> > > > * Parse the query params and re-construct the link such that your queries end up attached to the `link` query param.
> > > 
> > > 
> > > That's not always gonna work, sadly, if you are following an external redirect (like in most auth flows).
> > 
> > 
> > Isn't this the job of the external redirector though? e.g. Configure the authentication provider to keep query params.
> > I don't think this is related to either the OP or even dynamic links in general.
> 
> No, it is the job of dynamic links to not lose incoming query parameters.
> 
> Dynamic Link
> https://monadeo.page.link/github_auth?code=xxxxx
> 
> Converts to:
> 
> Observed:
> https://monadeo-test.firebaseapp.com/__/auth/handler
> 
> Expected:
> https://monadeo-test.firebaseapp.com/__/auth/handler?code=xxxxx
> 
> Try yourself to implement Github authentication for example and you’ll see that it cannot work because of the above.

Yes Dynamic links failed to send query params. This is one the most awaiting feature for us 😕

**Comment by fodil-a** [comment]
Any update on this please ? I still have this issue...

I am trying to get the email verification to work with a dynamic link to come back to app, especially by getting "oobCode" from the linkData, but it is not present

**Comment by RodrigoBertotti** [comment]
This worked for me:

https://**\<PREFIX\>**.page.link/?link=**[<ENCODED_URL>](https://www.urlencoder.org/)**&apn=**<MY_FLUTTER_APPLICATION_ID>**

Example:

https\://example.com/foo?boo=2

https://**my**.page.link/?link=**https%3A%2F%2Fexample.com%2Ffoo%3Fboo%3D2**&apn=**com.example.myflutterapplicationid**

Thanks to [Ashutosh Pathak](https://stackoverflow.com/a/62397843/4508758)

**Comment by astronaute77** [comment]
> 
> 
> This worked for me:
> 
> https://**<PREFIX>**.page.link/?link=**[<ENCODED_URL>](https://www.urlencoder.org/)**&apn=**<MY_FLUTTER_APPLICATION_ID>**
> 
> Example:
> 
> https://example.com/foo?boo=2
> 
> https://**my**.page.link/?link=**https%3A%2F%2Fexample.com%2Ffoo%3Fboo%3D2**&apn=**com.example.myflutterapplicationid**
> 
> Thanks to [Ashutosh Pathak](https://stackoverflow.com/a/62397843/4508758)

Guys, please stop posting solutions to a different problem. 
It’s great that you are trying to help, but you completely misunderstood the issue.

The problem will be solved when this URL:
https://monadeo.page.link/github_auth?code=xxxxx

will convert to:
https://monadeo-test.firebaseapp.com/__/auth/handler?code=xxxxx

**Comment by Salakar** [comment]
We would welcome a documentation PR for this. Would someone be willing to send one and tag me in it? Thanks

**Comment by radiKal07** [comment]
The problem is inside Firebase REST API, but can be fixed from the plugin as well.
The plugin does this REST call internally:

```
curl --request POST \
  --url 'https://firebasedynamiclinks.googleapis.com/v1/reopenAttribution?key={firebase_key}' \
  --header 'content-type: application/json' \
  --header 'x-ios-bundle-identifier: {bundle_identifier}' \
  --data '{
	"bundle_id": "{bundle_identifier}",
	"sdk_version": "4.0.8",
	"requestedLink": "https://subdomain.domain.com/path?param=123456"
}'
```
This call returns:
```
{
  "deepLink": "https://subdomain.domain.com/path",
  "resolvedLink": "https://subdomain.domain.com/?link=https://subdomain.domain.com/path&apn={bundle_identifier}&isi=1493910790&ibi={bundle_identifier}&st={st}&sd={sd}?&si={si}"
}
```
Then the URL sent back to Flutter is the URL got from the response from the `deepLink` field.
Notice how in the request we sent `https://subdomain.domain.com/path?param=123456` and in the response we got back `https://subdomain.domain.com/path`. So Firebase backend removes `?param=123456`. What can the Flutter plugin do is ignore the deepLink from response and send back to Dart the original requestedLink which also has the params.

This is what I noticed for iOS.
On Android the URL does not have the query parameters from the start, so I don't think we can make it work just from the Flutter plugin.

**Comment by Vasault** [comment]
Has this been fixed yet? branch.io actually supports params without going for the long url, i had to jump from firebase to branch, and now branch is having issues so i'm trying to go back to firebase once more

**Comment by bmennear** [comment]
Is there any plans for a solution to this? With the main issue being closed it appears not. This is basic functionality that should exist. Without this short links are completely unusable in our use case (web users on mobile or desktop devices). We need to append something that uniquely identifies a user like an id. https://short.link/QQQQ?id=1234 should not dump the query params on redirect. 

I tried using anchor links instead, which do get passed (https://short.link/QQQQ#test or even https://short.link/QQQQ/#/test) however mobile browsers drop the anchor link / fragment. Works fine on web but this was a hack to try and restore functionality that should exist out of the box.

**Comment by dtheodor79** [comment]
Sorry, is there any update on this issue? 

Unless I miss something, in our case where we use dynamic links for email password reset, the oobCode, mode and apiKey params are removed!

**Reply by ndesamuelmbah** [reply]
Still facing this issue. Has anyone found any solution yet?

**Reply by sarinupreti2021** [reply]
having the same issue. 

this is my dynamic links :

https://example.page.link/resetpassword?token=FDOJ@#$#$#$#$


when I try to parse this in the app it shows me :
https://example.page.link/resetpassword and queryParameters is always false.

**Reply by federicogomez1-gj** [reply]
Hello @sarinupreti2021.  Has found any solution yet?

**Comment by chunlee-thong** [comment]
You can use another package to handle the link like https://pub.dev/packages/uni_links or https://pub.dev/packages/app_links because firebase dynamic is just a place to register a domain and link for your app. in the end, it's just an app link or universal link. I've been tested and it works.

**Reply by noah-swf** [reply]
I can highly recommend [uni_links](https://pub.dev/packages/uni_links) for this problem👍Quite easy to use and does what it's supposed to do.

**Comment by ekox** [comment]
Same problem, and finally i got this conclusion :
- uni_links/ app_links not working on android development (localhost) because custom scheme deep link like "app://myapp?token=xyz...." will be rejected from chrome browser
- app link (https only) not working on development
- stay on firebase flutter with some change on auth flow

there are 2 options with firebase dynamic links : 
1. Using REST API Dynamic Link Firebase (POST) to send custom query param URL inside our server side (backend). After sending the request, we would get a dynamic link from the server, then forward or redirect to that link so our flutter app can parse all query param we sent to API before. 
pros : simple
cons : maybe not to secure because we sent our token to firebase

2. Create unique ID from flutter (device id + timestamp or whatever...), generate link (short/long and dont forget to save inside sharedpreference or whatever) , send that unique ID + generate link to our auth flow (Auth Web App). Our backend have to  save that unique ID and then link from flutter. After auth flow have finished, redirect to that generate link. Our flutter will be triggered to check that dynamic link, and GET HTTP API to our backend (with unique ID before) to GET our token.
pros : maybe more secure than first option
cons : change our auth flow in the end

Good luck my friend....(sorry my english is very bad...)

**Comment by LeonardoRosaa** [comment]
I suppose that problem occurs by the way the dynamic link is built.

The Firebase Dynamic Link requires Dynamic Link parameters to configure your query parameters in your dynamic link. I will highlight the Dynamic Link format that does work for me below:

```bash
https://your.subdomain.page.link/?link=your_deep_link&apn=android.package.name&ibi=ios.package.name&ius=your.custom.url.scheme
```

The link above should work well on the **Android** platform. However, for the **IOS** platform there’re two more tricks:

- Query parameters in the deep link should be encodable;
- It’s required the **ius** optional Dynamic Link parameter. You can click [here](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app) to look at how to configure your custom URL scheme and use **ius** parameter.

**Comment by JosteveGit** [comment]
There is a `getDynamicLink` method that expects a `Uri` to be passed. It returns a `PendingDynamicLinkData` and the `queryParameters` can be gotten from the response.

See this:

```
      final x = await FirebaseDynamicLinks.instance.getDynamicLink(
        Uri.parse(link),
      );
      final deepLink = x?.link;
      if (deepLink == null) {
        return;
      }
      final queryParameters = deepLink.queryParameters;
      final longLink = deepLink.toString();
```

**Comment by nasir-azeemi** [comment]
Hey guys, I am having this issue too. 
I have two dynamic links.
I tried to create both short and long links however, the full link is not returned by messging.getInitialLink()



Here's my code:
<img width="1219" alt="Screenshot 2022-12-05 at 22 20 23" src="https://user-images.githubusercontent.com/54790791/205701409-d10707f3-9eab-45b5-a44f-5f8d19e3cb11.png">

Here's my console logs:
<img width="682" alt="Screenshot 2022-12-05 at 21 39 56" src="https://user-images.githubusercontent.com/54790791/205701498-8dda4027-86e4-42e1-96a2-0448d5a50572.png">
<img width="682" alt="Screenshot 2022-12-05 at 21 51 34" src="https://user-images.githubusercontent.com/54790791/205701505-9230aa2a-881e-4cdf-9a0c-ea90b6d3edb0.png">

**Comment by SMKH-PRO** [comment]
So it was never solved?

**Comment by sebastijandumancic** [comment]
It seems like this still hasn't been solved. I haven't managed to pass a custom property to a deeplink to be resolved in-app, such as activation code / coupon / etc.

**Comment by sebastijandumancic** [comment]
Ok, I figured it out, and it should be a final solution to this. 

You definitely need to follow manually creating a dynamic link, but it doesn't tell you that you NEED to implement some tags in order to open up the correct app since it's going around what's in the Firebase Dynamic Link dashboard. 

So, based on this: https://firebase.google.com/docs/dynamic-links/create-manually, I've constructed a following link: 

`https://example.page.link/?link=https://www.example.com/id?id=12345&apn=com.example.app&amv=3&ibi=com.example.app&isi=1234567&ius=exampleapp`

Where you must specify correct bundle id to open up the app, I'm not sure about the version (amv) or isi, but ibi and ius is required. After that when you resolve the link in the navigator, I get: 

` {"minimumAppVersion": 3, "url": "https://www.setfords.co.uk/reference-id?id=12345", "utmParameters": {}}`

From where it's easy to resolve the id query param in the URL prop. Note that custom part is passed to the link parameter in the link, not in the initial path, since that's hardcoded in the firebase URL. You probably do need one base dynamic link defined in the dashboard so you get proper page.link web generated. 

Let me know if this works, it's been a journey.

**Reply by szaboagoston** [reply]
> Ok, I figured it out, and it should be a final solution to this.
> 
> You definitely need to follow manually creating a dynamic link, but it doesn't tell you that you NEED to implement some tags in order to open up the correct app since it's going around what's in the Firebase Dynamic Link dashboard.
> 
> So, based on this: https://firebase.google.com/docs/dynamic-links/create-manually, I've constructed a following link:
> 
> `https://example.page.link/?link=https://www.example.com/id?id=12345&apn=com.example.app&amv=3&ibi=com.example.app&isi=1234567&ius=exampleapp`
> 
> Where you must specify correct bundle id to open up the app, I'm not sure about the version (amv) or isi, but ibi and ius is required. After that when you resolve the link in the navigator, I get:
> 
> ` {"minimumAppVersion": 3, "url": "https://www.setfords.co.uk/reference-id?id=12345", "utmParameters": {}}`
> 
> From where it's easy to resolve the id query param in the URL prop. Note that custom part is passed to the link parameter in the link, not in the initial path, since that's hardcoded in the firebase URL. You probably do need one base dynamic link defined in the dashboard so you get proper page.link web generated.
> 
> Let me know if this works, it's been a journey.

For me it did not work either.

**Reply by mt-akar** [reply]
This worked for me like a charm. Instead of adding custom query parameters to the entire link, add them to the link passed to the `link` query parameter.

In my case, no other parameter, such as `ibi` and `isi` was required.

**Comment by victorhuberta** [comment]
Hi everyone, I am following the Firebase instructions for Sign-in with Email Link (`firebase_auth: ^4.6.3`, `firebase_dynamic_links: ^5.3.4`) and the URL encoding issue described by @kostacasa still remains. It looks similar to this, using the Firebase emulator and manually created dynamic link -- passed to url parameter of `ActionCodeSettings`: 

`http://127.0.0.1:9099/emulator/action?mode=signIn&lang=en&oobCode=-_FXnuwyj0--9hiJLSezy3hW68tGiba2i7hbqu7C5bLIq_-gh8jRvx&apiKey=fake-api-key&continueUrl=https%3A%2F%2Fmy.page.link%2F%3Fapn%3Dcom.my.client%26efr%3D1%26link%3Dhttps%3A%2F%2Fexample.com`

which resolves to...

`https://my.page.link/?apn=com.my.client&efr=1&link=https%3A%2F%2Fexample.com&mode=signIn&lang=en&oobCode=-_FXnuwyj0--9hiJLSezy3hW68tGiba2i7hbqu7C5bLIq_-gh8jRvx&apiKey=fake-api-key`

The part starting from &mode=signIn becomes parameters of my.page.link and not of https://example.com, which means my app received the deep link only as 'https://example.com' and so `Firebase.instance.isSignInWithEmailLink` will always fail.

**Comment by chanonsut** [comment]
Hey guys, if you guys are still stuck. I have this solution. it so basically but work.
1. You click a "Link detail" in firebase dynamic link. then you copy "Long Dynamic Link"
2. You will get link like this "https://app.page.link/?link=http://example.page.app&apn=com.example.app&isi=1111111111&ibi=com.example.app"
3. focus on parameter "link=http://example.page.app" in this you add variable in get method like "http://example.page.app?data=123". => example with full link : "https://app.page.link/?link=http://example.page.app?data=123&apn=com.example.app&isi=1111111111&ibi=com.example.app"
4. if you will use in flutter project try to use [firebase_dynamic_links](https://pub.dev/packages/firebase_dynamic_links)

example
```
  void fetchLinkData() async {
    try{
      var link = await FirebaseDynamicLinks.instance.getInitialLink();
      handleLinkData(link!);

    } catch (e) {
      print(e.toString());
    }
  }

  void handleLinkData(PendingDynamicLinkData data) async {
    final Uri? uri = data.link;
    print(uri) // => http://example.page.app?data=123
   }
```
I holp it can help you guys.

**Comment by jaymin2609** [comment]
Try Below the mentioned solution. It's worked for me

FirebaseDynamicLinks dynamicLinks = FirebaseDynamicLinks.instance;

          PendingDynamicLinkData? pendingDynamicLinkData =
              await dynamicLinks.getDynamicLink(Uri.parse("<YOUR_DYNAMIC_LNK>"));
          if (pendingDynamicLinkData != null &&
              pendingDynamicLinkData.link.hasQuery) {
            final queryParametersAll =
                pendingDynamicLinkData.link.queryParametersAll;
           
          }
